Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,702 @@
+
+This means we need to write an interpreter, so let's do just that.
+Remember that `Free` has two constructors, `Pure` and `Free`, so we need to pattern-match on those:
+
+```haskell
+runState :: State s a -> s -> (a, s)
+runState (Pure x) s = (x, s)
+runState (Free f) s =
+  let (m, s') = runStateF f s
+  in runState m s'
+```
+
+If you're familiar with the usual state monad, you may have noticed that we've essentially moved the implementation of `>>=` to `runState`.
+The free monad doesn't specify the meaning of monadic actions, so we have to decide what those actions mean when we're running it.
+
+To illustrate this point, we're going to write another interpreter, which is essentially a pretty-printer:
+
+```haskell
+printState :: (Show s, Show a) => State s a -> s -> String
+printState (Pure x) s = "pure (" <> show x <> "," <> show s <> ")"
+printState (Free m) s =
+  let (x, s') = runStateF m s
+  in "state change " <> show s <> " -> "<> show s' <> "\n"
+    <> printState x s'
+```
+
+If we run `someComputateion` defined above through this interpreter, we'll get the following output:
+
+```
+state change 1 -> 1
+state change 1 -> 2
+pure ((),2)
+```
+
+Let's summarize what we've learned so far.
+We can take any usual base functor for some monad, and get a `Monad` instance "for free".
+The caveat is that we need to define the semantics of this monad separately by writing an interpreter.
+
+### List as a free monad
+
+Not all free monads behave exactly the same as their regular counterparts.
+For instance, you might know that the list monad encodes non-determinism.
+However, if we derive a free monad for lists, the behavior is slightly different.
+Consider:
+
+```haskell
+listComputation :: FreeList Int
+listComputation = do
+  x <- liftF [1, 2, 3]
+  y <- liftF [70, 80]
+  z <- liftF [500, 600]
+  pure $ x+y+z
+
+printFreeList :: Show a => FreeList a -> String
+printFreeList (Pure x) = show x
+printFreeList (Free f) = "["
+  <> intercalate "," (printFreeList <$> f)
+  <> "]"
+```
+
+if we run `printFreeList listComputation`, we will get
+
+```haskell
+[[[571,671],[581,681]],[[572,672],[582,682]],[[573,673],[583,683]]]
+```
+
+Notice how it gets us what is essentially nested lists, unlike the regular list monad.
+Of course, we can get the standard list monad behavior by concatenating the nested lists, but we might not necessarily want to do that.
+
+In general, we can restore any proper monad behavior from a free monad, since it doesn't actually define any semantics, beyond those necessary for any monad.
+`free` provides a function for that:
+
+```haskell
+foldFree :: Monad m => (forall x. f x -> m x) -> Free f a -> m a
+```
+
+Given a function[^natural-transformation] converting `f x` into some monad `m x` for any `x`, we can convert `Free f a` into `m a`.
+Returning to our list example, we can get the original list behavior by applying `foldFree id`.
```

Comment:
Perhaps it's obvious, but it could be convenient to have the original behaviour right here (just for presentation purposes)